_posts/2019-09-30-bert-r/bert_r.Rmd

```diff
@@ -0,0 +1,536 @@
+
+There are several methods to install keras-bert in Python. 
+
+- in Jupyter Notebook, run:
+
+```{r eval=FALSE}
+!pip install keras-bert
+```
+
+- in Terminal (Linux, Mac OS), run:
+
+```{r eval=FALSE}
+python3 -m pip install keras-bert
+```
+
+- in Anaconda prompt (Windows), run:
+
+```{r eval=FALSE}
+conda install keras-bert
+```
+
+After this procedure, you can check whether keras-bert is installed or not.
+
+```{r eval=F}
+reticulate::py_module_available('keras_bert')
+```
+
+```{r eval=F}
+[1] TRUE
+```
+
+Finally, TensorFlow version should be 1.14/1.15. You can check it in the following form:
+
+```{r eval=F}
+tensorflow::tf_version()
+```
+
+```{r eval=F}
+[1] ‘1.14’
+```
+
+In a nutshell:
+
+```{r eval=F}
+pip install keras-bert
+pip install tensorflow==1.15
+```
+
+## What is BERT?
+
+BERT^[[BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding](https://arxiv.org/pdf/1810.04805.pdf)] is a pre-trained deep learning model introduced by Google AI Research which has been trained on Wikipedia and BooksCorpus. It has a unique way to understand the structure of a given text. Instead of reading the text from left to right or from right to left, BERT reads the entire word sequences at once. So, it allows to understanding a word based on its surroundings. There are different kind of pre-trained BERT models but the main difference between them is trained parameters. In our case, [BERT with 12-layers, 768-hidden, 12-heads, and 110M parameters](https://storage.googleapis.com/bert_models/2018_10_18/uncased_L-12_H-768_A-12.zip) will be used to create a text classification model.
+
+## Model structure
+
+Loading a pre-trained BERT model is straightforward. [Downloaded zip file contains](https://storage.googleapis.com/bert_models/2018_10_18/uncased_L-12_H-768_A-12.zip):
+
+- *bert_model.ckpt* which is for loading the weights from the TensorFlow checkpoint
+- *bert_config.json* which is a configuration file
+- *vocab.txt* which is for text tokenization
+
+```{r}
+pretrained_path = '/Users/turgutabdullayev/Downloads/uncased_L-12_H-768_A-12'
+config_path = file.path(pretrained_path, 'bert_config.json')
+checkpoint_path = file.path(pretrained_path, 'bert_model.ckpt')
+vocab_path = file.path(pretrained_path, 'vocab.txt')
+```
+
+## Import Keras-Bert module via reticulate
+
+Let's upload keras-bert via reticulate and prepare tokenizer object. BERT tokenizer will help us to turn words into indices.
+
+```{r}
+k_bert = import('keras_bert')
+token_dict = k_bert$load_vocabulary(vocab_path)
+tokenizer = k_bert$Tokenizer(token_dict)
+```
+
+### How tokenizer works?
+
+BERT uses a WordPiece tokenization strategy. If a word is Out-of-vocabulary (OOV), then BERT will break it down into subwords. (eating => eat, ##ing).
```

Comment:
As the subsequent figure 1 is about embeddings, I'd add another (short) paragraph about what is going on with embeddings here.